### 1. Write programs to process the above corpus and find answers to the following questions:

#### 1.1 Which nouns are more common in their plural form, rather than their singular form? (Only consider regular plurals, formed with the –s suffix – don't worry about irregular plurals for now.)

In [1]:
#1.1
import pyconll
directory = "/Users/shristidhakal/Documents/Grad School/Comp Linguistics/week 10/UD_English-GUM-master/en_gum-ud-train.conllu"
train_data = pyconll.load_from_file(directory)
for sentence in train_data:
    for token in sentence:
        if token.upos == 'NOUN' and token.xpos == 'NNS':
            print(token.form, token.upos, token.xpos) #displaying nouns in plural form

#below are the nouns more common in their plural form

Insights NOUN NNS
people NOUN NNS
elements NOUN NNS
artworks NOUN NNS
labels NOUN NNS
people NOUN NNS
artworks NOUN NNS
issues NOUN NNS
ways NOUN NNS
paintings NOUN NNS
surroundings NOUN NNS
studies NOUN NNS
works NOUN NNS
distinctions NOUN NNS
topics NOUN NNS
collections NOUN NNS
themes NOUN NNS
insights NOUN NNS
techniques NOUN NNS
processes NOUN NNS
artworks NOUN NNS
effects NOUN NNS
artworks NOUN NNS
implications NOUN NNS
techniques NOUN NNS
behaviours NOUN NNS
strengths NOUN NNS
humanities NOUN NNS
studies NOUN NNS
techniques NOUN NNS
reactions NOUN NNS
representations NOUN NNS
artworks NOUN NNS
processes NOUN NNS
fixations NOUN NNS
movements NOUN NNS
saccades NOUN NNS
saccades NOUN NNS
goals NOUN NNS
properties NOUN NNS
elements NOUN NNS
movements NOUN NNS
participants NOUN NNS
beliefs NOUN NNS
memories NOUN NNS
impressions NOUN NNS
ways NOUN NNS
arts NOUN NNS
factors NOUN NNS
Sons NOUN NNS
years NOUN NNS
surroundings NOUN NNS
aspects NOUN NNS
collections NOUN NNS
artworks NOUN N

chemicals NOUN NNS
results NOUN NNS
households NOUN NNS
households NOUN NNS
sectors NOUN NNS
uses NOUN NNS
appliances NOUN NNS
homes NOUN NNS
sectors NOUN NNS
factors NOUN NNS
types NOUN NNS
perks NOUN NNS
perils NOUN NNS
people NOUN NNS
CFLs NOUN NNS
sources NOUN NNS
CFLs NOUN NNS
people NOUN NNS
emissions NOUN NNS
CFLs NOUN NNS
products NOUN NNS
wastes NOUN NNS
effects NOUN NNS
chemicals NOUN NNS
effects NOUN NNS
responses NOUN NNS
reactions NOUN NNS
reactions NOUN NNS
types NOUN NNS
lights NOUN NNS
sources NOUN NNS
proponents NOUN NNS
CFLs NOUN NNS
bulbs NOUN NNS
terms NOUN NNS
materials NOUN NNS
proponents NOUN NNS
Operators NOUN NNS
Systems NOUN NNS
operators NOUN NNS
Systems NOUN NNS
operators NOUN NNS
mutants NOUN NNS
operators NOUN NNS
applications NOUN NNS
Keywords NOUN NNS
operators NOUN NNS
systems NOUN NNS
errors NOUN NNS
cases NOUN NNS
mutants NOUN NNS
operators NOUN NNS
operators NOUN NNS
technologies NOUN NNS
languages NOUN NNS
paradigms NOUN NNS
operators NOUN NNS
featu

people NOUN NNS
Topics NOUN NNS
stories NOUN NNS
theories NOUN NNS
topics NOUN NNS
guests NOUN NNS
ventures NOUN NNS
toys NOUN NNS
toys NOUN NNS
appearances NOUN NNS
dogs NOUN NNS
items NOUN NNS
quotes NOUN NNS
novels NOUN NNS
dogs NOUN NNS
videos NOUN NNS
Years NOUN NNS
Spouse(s) NOUN NNS
Children NOUN NNS
awards NOUN NNS
actresses NOUN NNS
roles NOUN NNS
films NOUN NNS
performances NOUN NNS
1960s NOUN NNS
films NOUN NNS
80s NOUN NNS
vacations NOUN NNS
houses NOUN NNS
parents NOUN NNS
years NOUN NNS
20s NOUN NNS
actresses NOUN NNS
films NOUN NNS
parts NOUN NNS
years NOUN NNS
others NOUN NNS
years NOUN NNS
roles NOUN NNS
others NOUN NNS
1950s NOUN NNS
films NOUN NNS
film-makers NOUN NNS
directors NOUN NNS
directors NOUN NNS
albums NOUN NNS
accomplishments NOUN NNS
founders NOUN NNS
problems NOUN NNS
studies NOUN NNS
years NOUN NNS
1980s NOUN NNS
lectures NOUN NNS
universities NOUN NNS
Theories NOUN NNS
fields NOUN NNS
works NOUN NNS
decades NOUN NNS
examples NOUN NNS
exercises NOUN NNS

chemicals NOUN NNS
spores NOUN NNS
concentrations NOUN NNS
spores NOUN NNS
petridishes NOUN NNS
spores NOUN NNS
results NOUN NNS
results NOUN NNS
colonies NOUN NNS
ants NOUN NNS
ants NOUN NNS
results NOUN NNS
species NOUN NNS
ants NOUN NNS
species NOUN NNS
lives NOUN NNS
years NOUN NNS
years NOUN NNS
tricks NOUN NNS
years NOUN NNS
dragons NOUN NNS
books NOUN NNS
books NOUN NNS
influences NOUN NNS
magicians NOUN NNS
magicians NOUN NNS
people NOUN NNS
magicians NOUN NNS
books NOUN NNS
senses NOUN NNS
props NOUN NNS
opponents NOUN NNS
books NOUN NNS
thoughts NOUN NNS
actions NOUN NNS
protests NOUN NNS
talks NOUN NNS
questions NOUN NNS
weapons NOUN NNS
things NOUN NNS
weapons NOUN NNS
systems NOUN NNS
arms NOUN NNS
journals NOUN NNS
missiles NOUN NNS
pipelines NOUN NNS
lanes NOUN NNS
supplies NOUN NNS
activities NOUN NNS
actions NOUN NNS
strikes NOUN NNS
countries NOUN NNS
signs NOUN NNS
strikes NOUN NNS
workers NOUN NNS
police NOUN NNS
things NOUN NNS
owners NOUN NNS
comments NOUN NNS
rea

events NOUN NNS
floods NOUN NNS
droughts NOUN NNS
temperatures NOUN NNS
weeds NOUN NNS
pests NOUN NNS
diseases NOUN NNS
findings NOUN NNS
yields NOUN NNS
causes NOUN NNS
steps NOUN NNS
miles NOUN NNS
emissions NOUN NNS
regulations NOUN NNS
deities NOUN NNS
worshippers NOUN NNS
sites NOUN NNS
sites NOUN NNS
worshippers NOUN NNS
deities NOUN NNS
gods NOUN NNS
religions NOUN NNS
aspects NOUN NNS
religions NOUN NNS
religions NOUN NNS
islands NOUN NNS
islands NOUN NNS
minutes NOUN NNS
islands NOUN NNS
km NOUN NNS
islands NOUN NNS
inhabitants NOUN NNS
outcrops NOUN NNS
islands NOUN NNS
species NOUN NNS
people NOUN NNS
islands NOUN NNS
settlers NOUN NNS
visitors NOUN NNS
days NOUN NNS
requirements NOUN NNS
controls NOUN NNS
months NOUN NNS
flights NOUN NNS
groups NOUN NNS
camps NOUN NNS
islands NOUN NNS
years NOUN NNS
passengers NOUN NNS
charges NOUN NNS
Vessels NOUN NNS
harbours NOUN NNS
businesses NOUN NNS
cars NOUN NNS
cars NOUN NNS
vans NOUN NNS
visitors NOUN NNS
locals NOUN NNS
columns N

bypasses NOUN NNS
streets NOUN NNS
avenues NOUN NNS
arterials NOUN NNS
avenues NOUN NNS
cities NOUN NNS
names NOUN NNS
colleges NOUN NNS
towns NOUN NNS
streets NOUN NNS
cities NOUN NNS
streets NOUN NNS
parts NOUN NNS
streets NOUN NNS
names NOUN NNS
streets NOUN NNS
names NOUN NNS
parts NOUN NNS
streets NOUN NNS
streets NOUN NNS
tracks NOUN NNS
bounds NOUN NNS
areas NOUN NNS
points NOUN NNS
Cities NOUN NNS
towns NOUN NNS
cities NOUN NNS
suburbs NOUN NNS
miles NOUN NNS
Riders NOUN NNS
contours NOUN NNS
depots NOUN NNS
riders NOUN NNS
hours NOUN NNS
times NOUN NNS
buildings NOUN NNS
structures NOUN NNS
cities NOUN NNS
excavations NOUN NNS
1970s NOUN NNS
days NOUN NNS
sights NOUN NNS
charms NOUN NNS
explorers NOUN NNS
Festivals NOUN NNS
festivals NOUN NNS
festivals NOUN NNS
children NOUN NNS
travellers NOUN NNS
parts NOUN NNS
motorways NOUN NNS
connections NOUN NNS
reaches NOUN NNS
hours NOUN NNS
connections NOUN NNS
roads NOUN NNS
hours NOUN NNS
mornings NOUN NNS
roads NOUN NNS
junctions 

spaces NOUN NNS
boxes NOUN NNS
boxes NOUN NNS
things NOUN NNS
books NOUN NNS
items NOUN NNS
boxes NOUN NNS
Photos NOUN NNS
contents NOUN NNS
boxes NOUN NNS
parachutes NOUN NNS
cases NOUN NNS
strategies NOUN NNS
Steps NOUN NNS
chutes NOUN NNS
arms NOUN NNS
arms NOUN NNS
arms NOUN NNS
sides NOUN NNS
G-forces NOUN NNS
arms NOUN NNS
arms NOUN NNS
chances NOUN NNS
chances NOUN NNS
lines NOUN NNS
obstructions NOUN NNS
companions NOUN NNS
skydivers NOUN NNS
chutes NOUN NNS
instructors NOUN NNS
parachutes NOUN NNS
steps NOUN NNS
Causes NOUN NNS
beliefs NOUN NNS
chances NOUN NNS
injuries NOUN NNS
times NOUN NNS
odds NOUN NNS
Tips NOUN NNS
efforts NOUN NNS
humans NOUN NNS
reasons NOUN NNS
strategies NOUN NNS
phases NOUN NNS
Warnings NOUN NNS
landings NOUN NNS
arms NOUN NNS
things NOUN NNS
Steps NOUN NNS
boxes NOUN NNS
activities NOUN NNS
bits NOUN NNS
items NOUN NNS
projects NOUN NNS
things NOUN NNS
things NOUN NNS
time-frames NOUN NNS
items NOUN NNS
items NOUN NNS
ones NOUN NNS
things NOUN NNS


### 1.2 Which word has the greatest number of distinct tags? What are the tags, and which parts of speech do they represent?

In [2]:
#1.2

words_list = []
tags_list = []
for sentence in train_data:
    for token in sentence:
        word = token.form
        tag = token.upos
        words_list.append(word)
        tags_list.append(tag)
        
dictionary = dict(zip(words_list, tags_list)) #creating a dictionary of the word and tag
words_nd_tags_lists = [(k, v) for k, v in dictionary.items()]
print(words_nd_tags_lists)

[('Aesthetic', 'ADJ'), ('Appreciation', 'NOUN'), ('and', 'CCONJ'), ('Spanish', 'PROPN'), ('Art', 'PROPN'), (':', 'PUNCT'), ('Insights', 'NOUN'), ('from', 'ADP'), ('Eye-Tracking', 'NOUN'), ('Claire', 'PROPN'), ('Bailey-Ross', 'PROPN'), ('claire.bailey-ross@port.ac.uk', 'PROPN'), ('University', 'PROPN'), ('of', 'ADP'), ('Portsmouth', 'PROPN'), (',', 'PUNCT'), ('United', 'PROPN'), ('Kingdom', 'PROPN'), ('Andrew', 'PROPN'), ('Beresford', 'PROPN'), ('a.m.beresford@durham.ac.uk', 'PROPN'), ('Durham', 'PROPN'), ('Daniel', 'PROPN'), ('Smith', 'PROPN'), ('daniel.smith2@durham.ac.uk', 'PROPN'), ('Warwick', 'PROPN'), ('c.l.h.warwick@durham.ac.uk', 'PROPN'), ('How', 'SCONJ'), ('do', 'VERB'), ('people', 'NOUN'), ('look', 'VERB'), ('at', 'ADP'), ('experience', 'NOUN'), ('art', 'NOUN'), ('?', 'PUNCT'), ('Which', 'PRON'), ('elements', 'NOUN'), ('specific', 'ADJ'), ('artworks', 'NOUN'), ('they', 'PRON'), ('focus', 'VERB'), ('on', 'ADP'), ('Do', 'VERB'), ('museum', 'NOUN'), ('labels', 'NOUN'), ('have', 

In [3]:
#1.2
import nltk
data = nltk.ConditionalFreqDist((word.lower(), tag) for (word, tag) in words_nd_tags_lists)
for word in sorted(data.conditions()):
    if len(data[word]) > 3:
        tags = [tag for (tag, _) in data[word].most_common()]
        print("Word with greater than 3 tags: " ,word, ' '.join(tags))
    

In [4]:
#1.2
for word in sorted(data.conditions()):
    if len(data[word]) > 2:
        tags = [tag for (tag, _) in data[word].most_common()]
        print("Word with greater than 2 tags: " , word, ' '.join(tags))

Word with greater than 2 tags:  as ADP SCONJ PROPN
Word with greater than 2 tags:  press NOUN PROPN VERB


In [5]:
#1.2
for word in sorted(data.conditions()):
    if len(data[word]) > 1:
        tags = [tag for (tag, _) in data[word].most_common()]
        print("Word with greater than 1 tag: " , word, ' '.join(tags))

Word with greater than 1 tag:  abstract ADJ NOUN
Word with greater than 1 tag:  academia NOUN PROPN
Word with greater than 1 tag:  act VERB PROPN
Word with greater than 1 tag:  ad X NOUN
Word with greater than 1 tag:  admiral NOUN PROPN
Word with greater than 1 tag:  advanced PROPN VERB
Word with greater than 1 tag:  after SCONJ ADP
Word with greater than 1 tag:  age PROPN NOUN
Word with greater than 1 tag:  agency NOUN PROPN
Word with greater than 1 tag:  agriculture PROPN NOUN
Word with greater than 1 tag:  aim VERB PROPN
Word with greater than 1 tag:  air NOUN PROPN
Word with greater than 1 tag:  airlines NOUN PROPN
Word with greater than 1 tag:  airport PROPN NOUN
Word with greater than 1 tag:  al PROPN X
Word with greater than 1 tag:  almanac PROPN NOUN
Word with greater than 1 tag:  alone ADJ PROPN
Word with greater than 1 tag:  alter PROPN VERB
Word with greater than 1 tag:  am SYM PROPN
Word with greater than 1 tag:  analysis NOUN PROPN
Word with greater than 1 tag:  angels NOU

Word with greater than 1 tag:  upon SCONJ ADP
Word with greater than 1 tag:  upper ADJ PROPN
Word with greater than 1 tag:  us PRON PROPN
Word with greater than 1 tag:  valley PROPN NOUN
Word with greater than 1 tag:  victory PROPN NOUN
Word with greater than 1 tag:  village NOUN PROPN
Word with greater than 1 tag:  virtual PROPN ADJ
Word with greater than 1 tag:  visitors NOUN PROPN
Word with greater than 1 tag:  visual ADJ PROPN
Word with greater than 1 tag:  voice NOUN PROPN
Word with greater than 1 tag:  wake NOUN VERB
Word with greater than 1 tag:  walking NOUN PROPN
Word with greater than 1 tag:  war PROPN NOUN
Word with greater than 1 tag:  wars PROPN NOUN
Word with greater than 1 tag:  waters PROPN NOUN
Word with greater than 1 tag:  wave PROPN NOUN
Word with greater than 1 tag:  wax NOUN PROPN
Word with greater than 1 tag:  ways NOUN PROPN
Word with greater than 1 tag:  wedding PROPN NOUN
Word with greater than 1 tag:  weekly ADJ PROPN
Word with greater than 1 tag:  welcome IN

Looking at the results above, there are no words with greater than 3 tags. There are two words "as" and "press" with greater than 2 tags, and there are many words with greater than 1 tag. Thus, the answer to 1.2 is the words "as" and "press" have the greatest number of distinct tags.

#### 1.3 List the tags in order of decreasing frequency. Which parts of speech do the 20 most frequent tags represent? Do you see any patterns?

In [6]:
#1.3
tag_fd = nltk.FreqDist(tag for (word, tag) in words_nd_tags_lists)
tag_fd.most_common(20) #displaying 20 most frequent parts of speech tags

[('NOUN', 4333),
 ('PROPN', 2587),
 ('VERB', 2507),
 ('ADJ', 1570),
 ('ADV', 479),
 ('NUM', 419),
 ('X', 132),
 ('ADP', 85),
 ('PRON', 63),
 ('AUX', 41),
 ('SCONJ', 40),
 ('DET', 28),
 ('INTJ', 26),
 ('PUNCT', 25),
 ('CCONJ', 17),
 ('SYM', 16),
 ('PART', 9)]

The most frequent POS tags results show that nouns, proper nouns, verbs, and adjective occur the most in the corpus with frequencies of 4333, 2587, 2507, and 1570 respectively. This is followed by adverbs, numeral, and others (full list above).

#### 1.4 Which tags are nouns most commonly found after, and which parts of speech do they represent?

In [7]:
#1.4
tags_bigram =nltk.bigrams(tag for (word,tag)  in  words_nd_tags_lists)

after_noun = nltk.FreqDist(tag1 for (tag1,tag2)  in  tags_bigram if tag2 == 'NOUN')

print (after_noun.most_common(20))

[('NOUN', 1739), ('VERB', 1010), ('ADJ', 724), ('PROPN', 435), ('ADV', 149), ('NUM', 125), ('X', 37), ('ADP', 33), ('PRON', 25), ('DET', 13), ('SCONJ', 12), ('INTJ', 7), ('PUNCT', 6), ('AUX', 6), ('SYM', 6), ('PART', 3), ('CCONJ', 3)]


NOUN, VERB, ADJ, PROPN, ADV, and NUM, among others shown above, are the parts of speech tags for which nouns are most commonly found after.

### 2. Generate some statistics for tagged data to answer the following questions:

#### 2.1 What proportion of word types are always assigned the same part-of-speech tag?

In [8]:
#2.1
data = nltk.ConditionalFreqDist((word.lower(), tag) for (word, tag) in words_nd_tags_lists)
single_tags_count = 0
nonsingle_tags_count = 0
for word in sorted(data.conditions()):
    if len(data[word]) == 1:
        single_tags = [tag for (tag, _) in data[word].most_common()]
        single_tags_count = single_tags_count + 1
    else:
        nonsingle_tags_count = nonsingle_tags_count + 1

proportion_single_tags = single_tags_count / (single_tags_count+nonsingle_tags_count)  
print("Proportion of word types that're assigned the same part-of-speech tag: ", proportion_single_tags)


Proportion of word types that're assigned the same part-of-speech tag:  0.9430784472665046


#### 2.2 How many words are ambiguous, in the sense that they appear with at least two tags?

In [9]:
#2.2
at_least_two_tags_ct = 0
other_words_count = 0
for word in sorted(data.conditions()):
    if len(data[word]) >= 2:
        at_least_two_tags = [tag for (tag, _) in data[word].most_common()]
        #print("Word with greater than 1 tag: " , word, ' '.join(at_least_two_tags))
        at_least_two_tags_ct = at_least_two_tags_ct + 1
    else:
        other_words_count = other_words_count + 1
print("No. of words that are ambiguous (at least with 2 tags): ",at_least_two_tags_ct)


No. of words that are ambiguous (at least with 2 tags):  632


#### 2.3 What percentage of word tokens in the UD Corpus involve these ambiguous words?

In [10]:
percent_ambiguous_words = at_least_two_tags_ct / (at_least_two_tags_ct+other_words_count)
print(percent_ambiguous_words, "percentage of word tokens in the UD Corpus involve these ambiguous words.")


0.05692155273349545 percentage of word tokens in the UD Corpus involve these ambiguous words.


### 3. Since you are used to the corpus, you will build a Most Frequent tagger. The most frequency tagger does not take context into account. It is a simple dictionary based tagger that assigns the most frequent tag to a word. You will get the word-tag counts from the downloaded corpora. You will test your dictionary based tagger on the test corpus available here https://github.com/UniversalDependencies/UD_English-GUM/blob/master/en_gum-ud-test.conllu:

#### 3.1 You will write a function that takes a sentence and the dictionary as input. The output should consist of a tag for each word in the sentence. Repeat the step for all the sentences in the test corpus.

#### The Most Frequent Tagger

In [11]:
#3.1
directory_train = "/Users/shristidhakal/Documents/Grad School/Comp Linguistics/week 10/UD_English-GUM-master/en_gum-ud-train.conllu"

train_data = pyconll.load_from_file(directory_train)
tags_list = []
tagged_words_list = []
for sentence in train_data:
    sentence_text = sentence.text
    tokenized_words = nltk.word_tokenize(sentence_text)
    tagged_words = nltk.pos_tag(tokenized_words)
    tagged_words_list.append(tagged_words)
    tags = [tag for (word, tag) in tagged_words]
    tags_list.append(tags)
most_freq_tag = nltk.FreqDist(tags).max()
most_freq_tag


'NNP'

In [12]:
#3.1
def most_frequent_tagger(directory_train):
    words_list = []
    tags_list = []
    for sentence in train_data:
        for token in sentence:
            word = token.form
            tag = token.upos
            words_list.append(word)
            tags_list.append(tag)
    
    dictionary = dict(zip(words_list, tags_list))    
    words_nd_tags_lists = [(k, v) for k, v in dictionary.items()]
    
    new_words_list = []
    new_tags_list = []
    for (word, tag) in words_nd_tags_lists:
        tagger = "NNP"
        word = word
        new_words_list.append(word)
        new_tags_list.append(tagger)
    
    dictionary = dict(zip(new_words_list, new_tags_list))    
    most_frequent_tag = [(k, v) for k, v in dictionary.items()]
    
    return most_frequent_tag

most_frequent_tagger(directory_train)


[('Aesthetic', 'NNP'),
 ('Appreciation', 'NNP'),
 ('and', 'NNP'),
 ('Spanish', 'NNP'),
 ('Art', 'NNP'),
 (':', 'NNP'),
 ('Insights', 'NNP'),
 ('from', 'NNP'),
 ('Eye-Tracking', 'NNP'),
 ('Claire', 'NNP'),
 ('Bailey-Ross', 'NNP'),
 ('claire.bailey-ross@port.ac.uk', 'NNP'),
 ('University', 'NNP'),
 ('of', 'NNP'),
 ('Portsmouth', 'NNP'),
 (',', 'NNP'),
 ('United', 'NNP'),
 ('Kingdom', 'NNP'),
 ('Andrew', 'NNP'),
 ('Beresford', 'NNP'),
 ('a.m.beresford@durham.ac.uk', 'NNP'),
 ('Durham', 'NNP'),
 ('Daniel', 'NNP'),
 ('Smith', 'NNP'),
 ('daniel.smith2@durham.ac.uk', 'NNP'),
 ('Warwick', 'NNP'),
 ('c.l.h.warwick@durham.ac.uk', 'NNP'),
 ('How', 'NNP'),
 ('do', 'NNP'),
 ('people', 'NNP'),
 ('look', 'NNP'),
 ('at', 'NNP'),
 ('experience', 'NNP'),
 ('art', 'NNP'),
 ('?', 'NNP'),
 ('Which', 'NNP'),
 ('elements', 'NNP'),
 ('specific', 'NNP'),
 ('artworks', 'NNP'),
 ('they', 'NNP'),
 ('focus', 'NNP'),
 ('on', 'NNP'),
 ('Do', 'NNP'),
 ('museum', 'NNP'),
 ('labels', 'NNP'),
 ('have', 'NNP'),
 ('an', '

#### 3.2 Now you need to evaluate the accuracy of the tagger that you just built.
1. Across all the sentences, how many words were tagged correctly? You can ignore
sentence distinction. What is the precision, recall, and F-score?
2. How many sentences have words that were tagged completely correct? You will need to
loop through each sentence to determine this count.

For 3.2, I didn't use the dictionary created in 3.1 because I ran into issues while printing the classification metrics. The metrics would not take in a dictionary for some reason. So, I used lists of lists instead of dictionary, then used the list of lists outputs (saved in most_frequent_tag variable) to identify correct/incorrect predictions and precision, recall, and F-score metrics.

In [13]:
#3.2.1

directory_test = "/Users/shristidhakal/Documents/Grad School/Comp Linguistics/week 10/UD_English-GUM-master/en_gum-ud-test.conllu"
test_data = pyconll.load_from_file(directory_test)

correct_pred_sentence_count = 0
tagged_words_list = []
gold_tags = []
pred_tags = []
sent_gold_nd_pred_tags_list = []

most_freq_tagger_list = []
for sentence in test_data:
    sentence_text = sentence.text
    tokenized_words = nltk.word_tokenize(sentence_text)
    tagged_words = nltk.pos_tag(tokenized_words)

    new_words_list = []
    new_tags_list = []

    for key, tagger in tagged_words:
        word = key
        tagger = "NNP"
        new_words_list.append(word)
        new_tags_list.append(tagger)

    most_frequent_tag = [(new_words_list[i],new_tags_list[i]) for i in range(0,len(new_words_list))]

    gold_sentences = nltk.pos_tag(tokenized_words) #words tagged with nltk POS tagger
    tagged_test_sentences = most_frequent_tag #words tagged with most frequent tagger
    
    for token, tag1 in gold_sentences:
        gold = str(tag1)
        gold_tags.append(gold)
            
    for token, tag2 in tagged_test_sentences:
        pred = str(tag2)
        pred_tags.append(pred)
    
    gold_nd_pred_tags = [(gold_tags[i],pred_tags[i]) for i in range(0,len(gold_tags))]
    for key, val in gold_nd_pred_tags:
        if key != val:
            count = 0
        else:
            count = 1
            correct_pred_sentence_count = correct_pred_sentence_count + count
print("No. of sentences with all tags correctly predicted: ", correct_pred_sentence_count)           


No. of sentences with all tags correctly predicted:  777128


This number looks ridiculously high. There might be a bug in the code, I couldn't figure out after trying a few times, so I'll just accept whatever credits you give me for this part.

In [14]:
#3.2.2
#Identifying across all the sentences, how many words were tagged correctly

gold_nd_pred_tags = [(gold_tags[i],pred_tags[i]) for i in range(0,len(gold_tags))]
count_correct = 0
for key, val in gold_nd_pred_tags:
    if key == val:
        count_correct = count_correct + 1
print("No. of words tagged correctly across all sentences: ", count_correct)


No. of words tagged correctly across all sentences:  1519


In [16]:
gold_tags

['DT',
 'NN',
 'IN',
 'NN',
 'IN',
 'JJ',
 'NNS',
 'IN',
 'JJ',
 'NNP',
 ':',
 'NNS',
 'IN',
 'DT',
 'RB',
 'JJ',
 'NN',
 'IN',
 'NNS',
 'NN',
 '.',
 'JJ',
 'NNS',
 'IN',
 'NN',
 'CC',
 'NNS',
 'VBP',
 'VBN',
 'DT',
 'NN',
 'IN',
 'JJ',
 'JJ',
 'NN',
 'NN',
 '.',
 'CC',
 'CD',
 ':',
 'CD',
 'NN',
 'NNS',
 'VBP',
 'VBN',
 'DT',
 'JJ',
 'NNS',
 'IN',
 'NN',
 'FW',
 'CD',
 '$',
 'CD',
 'NN',
 ':',
 'NNS',
 'VBP',
 'VBN',
 'DT',
 'JJ',
 'NNS',
 'WDT',
 'VBP',
 'JJ',
 'CC',
 'JJ',
 'NNS',
 'VBP',
 'CD',
 'NN',
 ':',
 'NNS',
 'VBP',
 'VBN',
 'DT',
 'JJ',
 'NNS',
 'IN',
 'NN',
 'FW',
 'CD',
 '$',
 'CD',
 'NN',
 ':',
 'CC',
 'JJ',
 'NNS',
 'VBP',
 'VBN',
 'DT',
 'JJ',
 'NNS',
 'IN',
 'JJ',
 'JJ',
 'JJ',
 'NNS',
 'VBN',
 'IN',
 'DT',
 'NN',
 'IN',
 'PRP$',
 'NNS',
 '.',
 '$',
 'CD',
 'NN',
 'IN',
 'JJ',
 'NNS',
 ',',
 'NNS',
 'RB',
 'VB',
 'DT',
 'NN',
 'IN',
 'NN',
 'VBG',
 'DT',
 'NNS',
 'CC',
 'NNS',
 'IN',
 'NN',
 'CC',
 'NN',
 '.',
 'CC',
 'CD',
 'NN',
 'WP',
 'RB',
 'MD',
 'RB',
 'VB',
 '

In [15]:
#3.1.1
from sklearn import metrics
print(metrics.classification_report(gold_tags, pred_tags))


              precision    recall  f1-score   support

           $       0.00      0.00      0.00         7
          ''       0.00      0.00      0.00        34
           (       0.00      0.00      0.00        85
           )       0.00      0.00      0.00        85
           ,       0.00      0.00      0.00       747
           .       0.00      0.00      0.00       761
           :       0.00      0.00      0.00       125
          CC       0.00      0.00      0.00       563
          CD       0.00      0.00      0.00       360
          DT       0.00      0.00      0.00      1396
          EX       0.00      0.00      0.00        25
          FW       0.00      0.00      0.00         5
          IN       0.00      0.00      0.00      1748
          JJ       0.00      0.00      0.00      1098
         JJR       0.00      0.00      0.00        38
         JJS       0.00      0.00      0.00        36
          MD       0.00      0.00      0.00       151
          NN       0.00    

/Users/shristidhakal/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/classification.py:1437: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


The accuracy of the most frequent tagger is 9%, which is very low. The precision, recall, and f-1 scores are all 0 for each parts-of-speech. It is clear that the tagger doesn't predict well.